txp/Pos/Binary/Txp/Core.hs

```diff
@@ -105,16 +108,21 @@ deriveSimpleBi ''T.TxAux [
     ]]
 
 instance Bi T.TxProof where
-  encode proof =  encodeListLen 3
-               <> encode (T.txpNumber proof)
-               <> encode (T.txpRoot proof)
-               <> encode (T.txpWitnessesHash proof)
-  decode = do
-    enforceSize "TxProof" 3
-    T.TxProof <$> decode <*>
-                  decode <*>
-                  decode
+    encode proof =  encodeListLen 3
+                 <> encode (T.txpNumber proof)
+                 <> encode (T.txpRoot proof)
+                 <> encode (T.txpWitnessesHash proof)
+    decode = do
+      enforceSize "TxProof" 3
+      T.TxProof <$> decode <*>
+                    decode <*>
+                    decode
 
 instance Bi T.TxPayload where
     encode T.UnsafeTxPayload {..} = encode $ zip (toList _txpTxs) _txpWitnesses
-    decode = T.mkTxPayload <$> decode
+    decode =
+        -- It's not possible to speed up serialization here as we lose
+        -- info on serialization and build it back when
+        -- deserializing. Without changing serialization protocol of
+        -- course.
```

Comment:
Yes we may want to talk about changing the serialized format of a `TxPayload`. The merkle tree itself ought to be serialized, rather than its `toList` representation which forgets all of the hashes.

There's one other option I could imagine, which doesn't require changing the format: if the decoder could get the bytes from which a term was parsed, then constructing the merkle tree wouldn't be quite as intensive: the bytes could be hashed directly, rather than (as is the case now) the term re-serialized and then hashed.

Reply:
> if the decoder could get the bytes from which a term was parsed

:(

This is hard with `cborg` afaik.

Reply:
I think it may be possible to do over top of cborg (or any incremental decoder).

```Haskell
data DecoderStep m t =
    Done !ByteString !ByteOffset t
  | Fail !ByteString !ByteOffset !Text
  | Partial (Maybe ByteString -> Decoder m t)

newtype Decoder m t = Decoder (m (DecoderStep m t))

withConsumedBytes :: Monad m => ByteString -> Decoder m t -> Decoder m (t, ByteString)
withConsumedBytes bs (Decoder decode) = Decoder (withConsumedBytes_ bs <$> decode)

withConsumedBytes_ :: Monad m => ByteString -> DecoderStep m t -> DecoderStep m (t, ByteString)
withConsumedBytes_ !bs (Done trailing offset t) = Done trailing offset (t, take offset bs)
withConsumedBytes_ _ (Fail trailing offset err) = Fail trailing offset err
withConsumedBytes_ !bs (Partial k) = Partial $ \mBs ->
  let bs' = maybe id (flip (<>)) mBs bs
  in  withConsumedBytes bs' (k mBs)
```

If it passes, you get the bytes and a guarantee that they're at least a well-formed serialization. By not forcing the first component `t`, it's possible that the verification step could be skipped (if the decoder definition is sufficiently lazy).